src/main/java/org/opensearch/security/action/onbehalf/CreateOnBehalfOfTokenAction.java

```diff
@@ -0,0 +1,180 @@
+                                RestStatus.SERVICE_UNAVAILABLE,
+                                "The OnBehalfOf token generating API has been disabled, see {link to doc} for more information on this feature." /* TODO: Update the link to the documentation website */
+                            )
+                        );
+                        return;
+                    }
+
+                    final String clusterIdentifier = clusterService.getClusterName().value();
+
+                    final Map<String, Object> requestBody = request.contentOrSourceParamParser().map();
+                    final String description = (String) requestBody.getOrDefault("description", null);
+
+                    final Integer tokenDuration = Optional.ofNullable(requestBody.get("durationSeconds"))
+                        .map(value -> (String) value)
+                        .map(Integer::parseInt)
+                        .map(value -> Math.min(value, OBO_MAX_EXPIRY_SECONDS)) // Max duration seconds are 600
+                        .orElse(OBO_DEFAULT_EXPIRY_SECONDS); // Fallback to default
+
+                    final Boolean roleSecurityMode = Optional.ofNullable(requestBody.get("roleSecurityMode"))
+                        .map(value -> (Boolean) value)
+                        .orElse(true); // Default to false if null
+
+                    final String service = (String) requestBody.getOrDefault("service", DEFAULT_SERVICE);
+                    final User user = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+                    Set<String> mappedRoles = mapRoles(user, /*Do not include host based mappings*/ null);
+
+                    builder.startObject();
+                    builder.field("user", user.getName());
+
+                    final String token = vendor.createJwt(
+                        clusterIdentifier,
+                        user.getName(),
+                        service,
+                        tokenDuration,
+                        mappedRoles.stream().collect(Collectors.toList()),
+                        user.getRoles().stream().collect(Collectors.toList()),
+                        roleSecurityMode
+                    );
+                    builder.field("authenticationToken", token);
+                    builder.field("durationSeconds", tokenDuration);
+                    builder.endObject();
+
+                    response = new BytesRestResponse(RestStatus.OK, builder);
+                } catch (final Exception exception) {
+                    builder.startObject().field("error", exception.toString()).endObject();
+
+                    response = new BytesRestResponse(RestStatus.INTERNAL_SERVER_ERROR, builder);
+                }
+                builder.close();
+                channel.sendResponse(response);
+            }
+        };
+    }
+
+    public Set<String> mapRoles(final User user, final TransportAddress caller) {
```

Comment:
nit: This method could be `private Set<String> mapRoles(final User user)` since the only place it is called would take `caller` in as null which means that host mapping won't be done on this endpoint.